Report per ut test

dougiechong opened the pull request:
<!--- Briefly describe your changes in the area below. Mention any issues this PR closes. -->
Changes the JUnitXML  reporter to report failures per test case instead of per assertion.

Closes #77.

## Description
<!--- If necessary, go into depth of what this pull request is doing. -->
Changes the JUnitXML  reporter to report failures per test case instead of per assertion.

## Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- Feel free to remove whole sections, not points within the sections, that do not apply -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->


- [x] **I am adding new or changing current functionality.**
  - [x] I have added or updated the tests for the new or changed functionality in `Tests/UnitTests`.
  - [x] I have added note(s) to `CHANGELOG.md` as necessary.

## Contributing

- [x] I have read and agree to the [Contributor Agreement](https://github.com/sassoftware/jsl-hamcrest/blob/master/ContributorAgreement.txt)

<!--- Replace your name and e-mail below -->
Signed-off-by: Douglas Chong <dougiechong@gmail.com>

dougiechong committed:
update to make multiple failures per test case
test cases are no longer numbered by assertion.
hash_results test cases are now lists of failures/successes.

dougiechong committed:
update test cases for new failure format
(one per test case)

dougiechong committed:
remove caching reporter, fails tests sometimes

dougiechong committed:
update changelog for junitxml reporting failures
per test case

dougiechong committed:
fix accidental removal of ).

dougiechong commented:
@emccorkle, @JustinChilton, please review when you get a chance

emccorkle commented on CHANGELOG.md:
```diff
@@ -2,6 +2,10 @@
 # Changelog
 All notable changes to this project will be documented in this file.
 
+## [1.3.0]
```
This should be underneath 1.2.0 as we haven't release 1.2.0 yet. Could you just rename that 1.2.0 heading to HEAD while you are at it please? In addition, could you refer to #87 and/or #77 in the changelog description?

emccorkle commented on Tests/UnitTests/Reporters/JunitXMLReporterTest.jsl:
```diff
@@ -47,7 +47,8 @@ setup_expr = Expr(
         // This allows us to write assertions that are reported as outside of any test.
         ::ut assert that(ut as expr(Name Expr(value)), matcher, label);
     );
-    report = Expr(Try(cached_report, cached_report = reporter << get report));
```
This was added to address flaky tests in #79. Multiple evals of `report` would give different XML since the `id` is based on the current time. We had tests that depended on multiple evals of `report` always giving the same XML.

dougiechong replied:
When I use the cached_report, I would get errors since the report would not update even between tests. The ids are checking using the regex id="\d+_\d+" so that it does not matter what time it is when the report is generated.

dougiechong replied:
I see what you mean by the flaky test, sometimes the second match is a little past the originally matched time.

emccorkle replied:
Yes, see [JunitXMLReporterTest.jsl#L165-L169](https://github.com/sassoftware/jsl-hamcrest/blob/f91cfdac472b4c83635cb97fdb4b1a510082cd3a/Tests/UnitTests/Reporters/JunitXMLReporterTest.jsl#L165-L169). L165 and L169 assume that the report XML is exactly the same. But without caching, those two reports have the possibility to be different.

Maybe add a `Teardown` to the `test case` to delete `cached_report`. Though it should be removed anyway since each test clears symbols and uses a fresh test case IIRC. Unless you ran `setup_expr` outside of the `test case` as symbols that exist when the test starts are not cleaned up.

dougiechong replied:
I will explicitly delete ``cached_report`` before making the expression for report to ensure it is not set.

dougiechong replied:
done in d89763bf14213a03beced1c8efc34f90241f8d24

dougiechong committed:
update changelog from 1.2.0 to HEAD
update to link JunitXMLReporter now reports failures per test case
intead of per assertion issue .

emccorkle commented on Tests/UnitTests/Reporters/JunitXMLReporterTest.jsl:
```diff
@@ -140,20 +141,19 @@ ut test(JunitXMLReporters, "Counts are correct at root", Expr(
 
 ut test( JunitXMLReporters, "JunitXMLReporter one simple passing", Expr(
-
+//include("../../../Source/Reporters/JunitXMLReporter.jsl");
```
Remove the dead code please. I hope that the issue you were trying to solve with this code are be addressed by #66 and #67. Try building and installing the JSL-Hamcrest-Dev addin if you haven't already. See [README.md#releasing-a-version](https://github.com/sassoftware/jsl-hamcrest/blob/master/README.md#releasing-a-version) for details.

dougiechong replied:
done in de325ccbad38866829e3e311823355d232c5b9fd

emccorkle commented on Source/Reporters/JunitXMLReporter.jsl:
```diff
@@ -159,13 +164,10 @@ Define Class(
-
-      // find test results that are non empty
-      num_successes = Nitems(loc(test_case_info, [=>]));
+      show(num_successes);
```
Remove `show` please.

dougiechong replied:
done in f91cfdac472b4c83635cb97fdb4b1a510082cd3a

emccorkle commented on Source/Reporters/JunitXMLReporter.jsl:
```diff
@@ -206,11 +208,11 @@ Define Class(
     Match(NItems(list_label),
       4,
       curr_suite = Trim(list_label[1]);
-      curr_test = Trim(list_label[2]) || "_" || Trim(list_label[3]) || "_" || Trim(list_label[4]);
```
Remove dead code please.

dougiechong replied:
done in f91cfdac472b4c83635cb97fdb4b1a510082cd3a

dougiechong committed:
remove dead code

emccorkle commented on Source/Reporters/JunitXMLReporter.jsl:
```diff
@@ -221,15 +222,15 @@ Define Class(
 
       // find number of test cases in anonymous_test_suite with form label_XX where XX is any number
-      curr_anonymous_test_cases = 0;
+      /*curr_anonymous_test_cases = 0;
```
More dead code here.

dougiechong replied:
done in de325ccbad38866829e3e311823355d232c5b9fd

dougiechong committed:
delete cached_report explicitly
This is done to ensure cahced_report is not defined somewhere else
before the first call.

dougiechong committed:
remove dead code

dougiechong commented:
@emccorkle, are any other changes required before merging?

emccorkle commented:
@dougiechong Justin and I have been busy with other projects but I'll start my review today. Sorry, I should have mentioned that it might be several days before we got to the substance of your PR. Are you feeding this XML into any downstream tools? Right now, I'm just seeing how Jenkins JUnit plugin handles the output.

dougiechong commented:
No worries, the jenkins JUnit plugin should be fine. We feed it into Gitlab's test runner.
